beacon-chain/p2p/types/types.go

```diff
@@ -0,0 +1,139 @@
+// Package types contains all the respective p2p types that are required for sync
+// but cannot be represented as a protobuf schema. This package also contains those
+// types associated fast ssz methods.
+package types
+
+import (
+	ssz "github.com/ferranbt/fastssz"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// SSZUint64 is a uint64 type that satisfies the fast-ssz interface.
+type SSZUint64 uint64
+
+const rootLength = 32
+
+const maxErrorLength = 256
+
+// MarshalSSZTo marshals the uint64 with the provided byte slice.
+func (s *SSZUint64) MarshalSSZTo(dst []byte) ([]byte, error) {
+	marshalledObj, err := s.MarshalSSZ()
+	if err != nil {
+		return nil, err
+	}
+	return append(dst, marshalledObj...), nil
+}
+
+// MarshalSSZ Marshals the uint64 type into the serialized object.
+func (s *SSZUint64) MarshalSSZ() ([]byte, error) {
+	marshalledObj := ssz.MarshalUint64([]byte{}, uint64(*s))
+	return marshalledObj, nil
+}
+
+// SizeSSZ returns the size of the serialized representation.
+func (s *SSZUint64) SizeSSZ() int {
+	return 8
+}
+
+// UnmarshalSSZ unmarshals the provided bytes buffer into the
+// uint64 object.
+func (s *SSZUint64) UnmarshalSSZ(buf []byte) error {
+	if len(buf) != s.SizeSSZ() {
+		return errors.Errorf("expected buffer with length of %d but received length %d", s.SizeSSZ(), len(buf))
+	}
+	*s = SSZUint64(ssz.UnmarshallUint64(buf))
+	return nil
+}
+
+// BeaconBlockByRootsReq specifies the block by roots request type.
+type BeaconBlockByRootsReq [][rootLength]byte
+
+// MarshalSSZTo marshals the block by roots request with the provided byte slice.
+func (s *BeaconBlockByRootsReq) MarshalSSZTo(dst []byte) ([]byte, error) {
+	marshalledObj, err := s.MarshalSSZ()
+	if err != nil {
+		return nil, err
+	}
+	return append(dst, marshalledObj...), nil
+}
+
+// MarshalSSZ Marshals the block by roots request type into the serialized object.
+func (s *BeaconBlockByRootsReq) MarshalSSZ() ([]byte, error) {
+	if len(*s) > int(params.BeaconNetworkConfig().MaxRequestBlocks) {
+		return nil, errors.Errorf("beacon block by roots request exceeds max size: %d > %d", len(*s), params.BeaconNetworkConfig().MaxRequestBlocks)
+	}
+	buf := make([]byte, 0, s.SizeSSZ())
+	for _, r := range *s {
```

Comment:
Do you have to deference the pointer of s?

Reply:
You cant loop through a pointer of a slice, which is why it needs to be dereferenced here.

Reply:
Makes sense thanks